src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -0,0 +1,365 @@
+
+            protected PrimaryKey min, max;
+            protected long tokenCount;
+
+            // iterator with the least number of items
+            protected RangeIterator minRange;
+            // iterator with the most number of items
+            protected RangeIterator maxRange;
+
+            // tracks if all of the added ranges overlap, which is useful in case of intersection,
+            // as it gives direct answer as to such iterator is going to produce any results.
+            private boolean isOverlapping = true;
+
+            public Statistics(IteratorType iteratorType)
+            {
+                this.iteratorType = iteratorType;
+            }
+
+            /**
+             * Update statistics information with the given range.
+             *
+             * Updates min/max of the combined range, token count and
+             * tracks range with the least/most number of tokens.
+             *
+             * @param range The range to update statistics with.
+             */
+            public void update(RangeIterator range)
+            {
+                switch (iteratorType)
+                {
+                    case CONCAT:
+                        // range iterators should be sorted, but previous max must not be greater than next min.
+                        if (range.getCount() > 0)
+                        {
+                            if (tokenCount == 0)
+                            {
+                                min = range.getMinimum();
+                            }
+                            else if (tokenCount > 0 && max.compareTo(range.getMinimum()) > 0)
+                            {
+                                throw new IllegalArgumentException("RangeIterator must be sorted, previous max: " + max + ", next min: " + range.getMinimum());
+                            }
+
+                            max = range.getMaximum();
+                        }
+                        break;
+
+                    case UNION:
+                        min = nullSafeMin(min, range.getMinimum());
+                        max = nullSafeMax(max, range.getMaximum());
+                        break;
+
+                    case INTERSECTION:
+                        // minimum of the intersection is the biggest minimum of individual iterators
+                        min = nullSafeMax(min, range.getMinimum());
+                        // maximum of the intersection is the smallest maximum of individual iterators
+                        max = nullSafeMin(max, range.getMaximum());
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Unknown iterator type: " + iteratorType);
+                }
+
+                // check if new range is disjoint with already added ranges, which means that this intersection
+                // is not going to produce any results, so we can cleanup range storage and never added anything to it.
```

Comment:
Aside: I kind of wonder why we didn't just calculate `isDisjoint`, given we always negate :)